src/metrics/metrics_provider.cpp

```diff
@@ -21,27 +21,103 @@
+{
+    QNetworkAccessManager manager;
+    QNetworkRequest request{metrics_url};
+    request.setHeader(QNetworkRequest::ContentTypeHeader, "application/json");
+    request.setHeader(QNetworkRequest::ContentLengthHeader, body.size());
+    request.setHeader(QNetworkRequest::UserAgentHeader, "multipassd/1.0");
+
+    QEventLoop event_loop;
+    std::unique_ptr<QNetworkReply> reply;
+
+    // For unit testing
+    if (metrics_url.isLocalFile())
+        reply = std::unique_ptr<QNetworkReply>(manager.put(request, body));
+    else
+        reply = std::unique_ptr<QNetworkReply>(manager.post(request, body));
+
+    QObject::connect(reply.get(), &QNetworkReply::finished, &event_loop, &QEventLoop::quit);
+
+    event_loop.exec();
+
+    auto buff = reply->readAll();
+
+    if (reply->error() == QNetworkReply::ProtocolInvalidOperationError)
+    {
+        auto error_msg = QJsonDocument::fromJson(buff).object();
+        mpl::log(mpl::Level::error, category,
+                 fmt::format("Metrics error: {} - {}", error_msg["code"].toString().toStdString(),
+                             error_msg["message"].toString().toStdString()));
+    }
+    else if (reply->error() != QNetworkReply::NoError)
+    {
+        QObject::disconnect(reply.get(), 0, 0, 0);
+        throw std::runtime_error(fmt::format("Metrics error: {}", reply->errorString().toStdString()));
+    }
+}
 }
 
-mp::MetricsProvider::MetricsProvider(const QUrl& metrics_url, const QString& unique_id)
-    : metrics_url{metrics_url}, unique_id{unique_id}
+mp::MetricsProvider::MetricsProvider(const QUrl& url, const QString& unique_id)
+    : metrics_url{url}, unique_id{unique_id}, metrics_sender{[this] {
+          std::unique_lock<std::mutex> lck(metrics_mutex, std::defer_lock);
+          while (true)
+          {
+              while (!lck.try_lock())
+                  std::this_thread::sleep_for(1ms);
+
+              if (metric_batches.empty())
+              {
+                  metrics_cv.wait(lck);
+              }
+
+              // Null element signals thread to exit
+              if (metric_batches.first().isNull())
+              {
+                  lck.unlock();
+                  return;
+              }
+
+              auto body = QJsonDocument(metric_batches).toJson(QJsonDocument::Compact);
+              lck.unlock();
+
+              try
+              {
+                  post_request(metrics_url, body);
+                  lck.lock();
+                  if (metric_batches.first().isNull())
+                  {
+                      lck.unlock();
+                      return;
+                  }
+                  metric_batches = QJsonArray{};
+                  lck.unlock();
+              }
+              catch (const std::exception& e)
+              {
+                  mpl::log(mpl::Level::error, category, fmt::format("{} - Attempting to resend", e.what()));
+                  std::this_thread::sleep_for(60s);
```

Comment:
I think this should probably be an exponentially backoff timeout.

Also instead of sleep (which cannot be aborted) maybe using the timeout version of the condition variable is best.

So maybe the loop looks something like:
```
std::unique_lock<std::mutex> lock{metrics_mutex};
auto timeout = std::chrono::seconds::max(); //max because we don't want to initially time out until metrics are available
while (running)
{
cv.wait_for(lock, timeout, [&] { return metrics_available || !running;})

if (!running) return;
if (!metrics_available) continue; //timed out and no metrics available yet

<attempt to send metrics>
if (failed)
   timeout = <increase timeout exponentially until some ceiling value?>;
else
  return;
}
```

Reply:
Yeah, I was planning on adding a backoff timeout for each network failure.  Also, we do want to unlock before sending the metrics to allow any new metrics to be added to the JsonArray independent of the thread trying to send out the metrics.  It allows for queuing of metrics by adding a new JSonArray entry.